src/theory/uf/distinct_extension.cpp

```diff
@@ -0,0 +1,252 @@
+          Assert(i < d_eqcToDMem[t1].size());
+          Assert(i2 < d_eqcToDMem[t2].size());
+          Node eq = d_eqcToDMem[t1][i].eqNode(d_eqcToDMem[t2][i2]);
+          Assert(d_state.areEqual(d_eqcToDMem[t1][i], d_eqcToDMem[t2][i2]));
+          Trace("uf-lazy-distinct")
+              << "...conflict " << eq << " " << d << std::endl;
+          std::vector<Node> exp{eq, d};
+          d_pendingConflict = nodeManager()->mkAnd(exp);
+          return;
+        }
+        Trace("uf-lazy-distinct") << "...no conflict" << std::endl;
+      }
+    }
+    else
+    {
+      d1.clear();
+      d1m.clear();
+    }
+    // append lists
+    d1.insert(d1.end(), d2.begin(), d2e);
+    std::vector<Node>& d2m = d_eqcToDMem[t2];
+    d1m.insert(d1m.end(), d2m.begin(), d2m.begin() + it2->second);
+  }
+}
+
+void DistinctExtension::check(Theory::Effort level)
+{
+  if (!d_pendingConflict.get().isNull())
+  {
+    Node conf = d_pendingConflict.get();
+    std::vector<Node> exp(conf.begin(), conf.end());
+    d_im.conflictExp(InferenceId::UF_DISTINCT_DEQ, exp, d_epg.get());
+    return;
+  }
+  if (level != Theory::Effort::EFFORT_LAST_CALL)
+  {
+    return;
+  }
+  TheoryModel* tm = d_state.getModel();
+  bool addedLemma = false;
+  // check negated distinct
+  size_t nnd = d_negDistinct.size();
+  for (size_t i = d_negDistinctIndex.get(); i < nnd; i++)
+  {
+    Node ndistinct = d_negDistinct[i];
+    Assert(ndistinct.getKind() == Kind::NOT
+           && ndistinct[0].getKind() == Kind::DISTINCT);
+    // check if satisfied
+    Node atom = ndistinct[0];
+    std::unordered_set<Node> reps;
+    bool isSat = false;
+    for (size_t j = 0, nterms = atom.getNumChildren(); j < nterms; j++)
+    {
+      Node ncr = d_state.getRepresentative(atom[j]);
+      if (!reps.insert(ncr).second)
+      {
+        isSat = true;
+        break;
+      }
+    }
+    if (!isSat)
+    {
+      Node batom = TheoryUfRewriter::blastDistinct(nodeManager(), atom);
+      Node lem = nodeManager()->mkNode(Kind::IMPLIES, batom, atom);
+      TrustNode tlem = TrustNode::mkTrustLemma(lem, nullptr);
+      if (d_im.trustedLemma(tlem, InferenceId::UF_NOT_DISTINCT_ELIM))
+      {
+        addedLemma = true;
+      }
+    }
+    // otherwise we determined that we are SAT in the remainder of this
+    // SAT context (due to looking up representatives, instead of model values)
+    // meaning that we don't need to check this constraint again in this
+    // SAT context.
+  }
+  d_negDistinctIndex = nnd;
+  if (addedLemma)
+  {
+    return;
+  }
+  // Note that it still may the case that we have a (distinct t1 ... tn)
+  // constraint where ti and tj were assigned the same value in the model,
+  // but the UF theory was not informed of ti = tj. This is because we
+  // do not insist that distinct induces care pairs in theory combination.
+  // Thus we must double check that are values are distinct in the model.
+  // If not, then we add the lemma (~distinct(t1,...,tn) or ti != tj).
```

Comment:
IIUC you are doing this for all cases at once, rather than for a single case, right? Please clarify.